assignment/impl/src/java/org/sakaiproject/assignment/impl/sort/AssignmentSubmissionComparator.java

```diff
@@ -62,11 +66,45 @@ public AssignmentSubmissionComparator(AssignmentService assignmentService, SiteS
 
     @Override
     public int compare(AssignmentSubmission a1, AssignmentSubmission a2) {
-        int result;
+        // Compare by UserSortNameComparator if appropriate
+        Optional<User> u1 = getIndividualSubmitterAsUser(a1);
+        Optional<User> u2 = getIndividualSubmitterAsUser(a2);
+        if (u1.isPresent() && u2.isPresent()) {
+            return userSortNameComparator.compare(u1.get(), u2.get());
+        }
+
         String name1 = getSubmitterSortname(a1);
```

Comment:
Do we need this now? if there is no user... getSubmitterSortname(a1) is empty?

Reply:
I assumed it was the group name or group id for a group submission? Or anon submission?

Reply:
The getSubmitterSortname method is implemented below, and it acts as a fallback for the following cases:

- the submission is associated with a group assignment
- the submission does not have exactly 1 associated submitter
- the submitter cannot be retrieved via the UserDirectoryService

I haven't actually tested anonymous assignments, but I didn't see any special handling in the existing code. Anonymous assignments should be using AnonymousSubmissionComparator from what I can tell

Reply:
For group submissions should we use the group name in this case?

Reply:
Right - if a submission is for a group assignment, getIndividualSubmitterAsUser returns early with an empty optional on line 91. Since the optionals won't be present on line 72, we fall back to comparing getSubmitterSortName(...), which uses group titles down on line 137:
`buffer.append(site.getGroup(assignmentService.getSubmissionSubmittee(_submission).get().getSubmitter()).getTitle());`